lib/edda-server/src/compressor.rs

```diff
@@ -0,0 +1,422 @@
+                                    }
+                                    // This should be impossible since compressing should always
+                                    // result in an active state.
+                                    State::Inactive => {
+                                        error!("this should be impossible");
+                                        break 'inner;
+                                    }
+                                };
+
+                                // Reset the state and send the compressed message to the naxum handler.
+                                // This is the end of the critical path!
+                                *project.state = State::Inactive;
+
+                                // FIXME(nick): I spent an hour trying to make naxum generic. Basically "Message<R>" is used everywhere
+                                // and "R" must implement "MessageHead". I don't think it is worth pursuing that path. Frankly, I am
+                                // not certain naxum is even the right tool for this. We should probably just double ack every message
+                                // here and switch to more primitive Rust stream management semantics.
+                                // ```
+                                // message.payload =
+                                //     Bytes::from(serde_json::to_value(new_message)?);
+                                // ```
+
+                                // Return the compressed message. This is the critical path!
+                                return Poll::Ready(Some(Ok(cloned_message)));
+                            }
+                            // Everything is not drained so let's update the state and LOOP again!
+                            // This is the critical path!
+                            Poll::Ready(Ok(state)) => {
+                                // FIXME(nick): double ack the message since it has been compressed.
+                                *project.state = state;
+                                break 'inner;
+                            }
+                            Poll::Ready(Err(err)) => {
+                                error!(si.error.message = ?err, "compress error");
+                                break 'inner;
+                            }
+                            // We are waiting for the layer db future.
+                            Poll::Pending => {}
+                        }
+                    }
+                }
+                // FIXME(nick): handle shutdown and publishing what we have compressed in flight on the way out.
+                Poll::Ready(None) => todo!(),
+                // FIXME(nick): handle publishing what we have compressed in flight before returning pending.
+                Poll::Pending => todo!(),
+            }
+        }
+    }
+}
+
+async fn compress(ctx: &DalContext, request: EddaRequestKind, state: State) -> Result<State> {
```

Comment:
The core business logic.

If the function body looks a bit strange, it is because `state` was originally passed in as a `&mut`. That was great for directly mutating the projection from `pin-project-lite` but WOW did things get confusing and likely error prone. Lot of `(*my_vec).push(*my_id)` code.